File: Assets/Script/GamePlay/Buff/BuffZombieSkin.cs
using UnityEngine;
using GamePlay.Units;
using Status.damage;

namespace GamePlay.Buff
{
    public class BuffZombieSkin : BuffBase
    {
        [SerializeField] private int damageReduction = 2;

        public BuffZombieSkin()
        {
            MaxStacks = 1;
            DecayAtTurnStart = false;
        }

        public override void OnApply(MapUnit owner)
        {
            base.OnApply(owner);
            Initialize(owner);
        }

        public override void OnIncomingDamage(ref float damage, DamageInfo damageInfo)
        {
            if (Owner == null) return;

            if (!Owner.hasMoved)
            {
                float reduction = Stacks * damageReduction;
                damage = Mathf.Max(0, damage - reduction);
            }
        }
    }
}

File: Assets/Lua/Buffs/BuffZombieSkin.lua
local BuffBase = require("Buffs.BuffBase")

---@class BuffZombieSkin : BuffBase
---@diagnostic disable-next-line: undefined-field
local BuffZombieSkin = BuffBase.extends(BuffBase)

function BuffZombieSkin:ctor(stacks)
    self.super.ctor(self, stacks or 1)
    self.Name = "僵尸皮肤"
    self.MaxStacks = 1
    self.DecayAtTurnStart = false
    self.DamageReduction = 2
end

function BuffZombieSkin:OnApply(owner) end

function BuffZombieSkin:OnTurnStart(owner) end

function BuffZombieSkin:OnIncomingDamage(damage, info)
    if self._Owner and not self._Owner.hasMoved then
        local reduction = self.Stacks * self.DamageReduction
        damage = math.max(0, damage - reduction)
    end
    return damage
end

return BuffZombieSkin
